README.rst
===========
python-nmap
===========

|PyPI latest| |PyPI Version| |PyPI License|

python-nmap is a python library which helps in using nmap port scanner. It allows to easilly manipulate nmap scan results and will be a perfect tool for systems administrators who want to automatize scanning task and reports. It also supports nmap script outputs.

It can even be used asynchronously. Results are returned one host at a time to a callback function defined by the user.

Download latest
===============

    python-nmap-0.4.1.tar.gz - 2015-08-21
    md5sum is b466e4b2ef30a0b9c0cb80aac215fb79

Warning : this version is intended to work with Python 3.x. For Python 2.x, please use python-nmap-0.1.4.tar.gz

Download development version
============================


.. code-block:: bash

    $ hg clone https://bitbucket.org/xael/python-nmap

Installation
============

From the shell, uncompress python-nmap-0.4.1.tar.gz and then run make :

.. code-block:: bash

    $ tar xvzf python-nmap-0.4.1.tar.gz
    $ cd python-nmap-0.4.1
    $ python setup.py install

or using Pip

.. code-block:: bash

    $ pip install python-nmap

Now you may invoke nmap from python


.. code-block:: bash

    $ python
    Python 2.6.4 (r264:75706, Dec  7 2009, 18:45:15)
    [GCC 4.4.1] on linux2
    Type "help", "copyright", "credits" or "license" for more information.
    >> import nmap


Usage
=====

From python/ipython:
--------------------

.. code-block:: python

    >>> import nmap
    >>> nm = nmap.PortScanner()
    >>> nm.scan('127.0.0.1', '22-443')
    >>> nm.command_line()
    'nmap -oX - -p 22-443 -sV 127.0.0.1'
    >>> nm.scaninfo()
    {'tcp': {'services': '22-443', 'method': 'connect'}}
    >>> nm.all_hosts()
    ['127.0.0.1']
    >>> nm['127.0.0.1'].hostname()
    'localhost'
    >>> nm['127.0.0.1'].state()
    'up'
    >>> nm['127.0.0.1'].all_protocols()
    ['tcp']
    >>> nm['127.0.0.1']['tcp'].keys()
    [80, 25, 443, 22, 111]
    >>> nm['127.0.0.1'].has_tcp(22)
    True
    >>> nm['127.0.0.1'].has_tcp(23)
    False
    >>> nm['127.0.0.1']['tcp'][22]
    {'state': 'open', 'reason': 'syn-ack', 'name': 'ssh'}
    >>> nm['127.0.0.1'].tcp(22)
    {'state': 'open', 'reason': 'syn-ack', 'name': 'ssh'}
    >>> nm['127.0.0.1']['tcp'][22]['state']
    'open'

    >>> for host in nm.all_hosts():
    >>>     print('----------------------------------------------------')
    >>>     print('Host : %s (%s)' % (host, nm[host].hostname()))
    >>>     print('State : %s' % nm[host].state())
    >>>     for proto in nm[host].all_protocols():
    >>>         print('----------')
    >>>         print('Protocol : %s' % proto)
    >>>
    >>>         lport = nm[host][proto].keys()
    >>>         lport.sort()
    >>>         for port in lport:
    >>>             print ('port : %s\tstate : %s' % (port, nm[host][proto][port]['state']))
    ----------------------------------------------------
    Host : 127.0.0.1 (localhost)
    State : up
    ----------
    Protocol : tcp
    port : 22	state : open
    port : 25	state : open
    port : 80	state : open
    port : 111	state : open
    port : 443	state : open


To export to a file
-------------------

.. code-block:: python

    >>> print(nm.csv())
    host;protocol;port;name;state;product;extrainfo;reason;version;conf
    127.0.0.1;tcp;22;ssh;open;OpenSSH;protocol 2.0;syn-ack;5.9p1 Debian 5ubuntu1;10
    127.0.0.1;tcp;25;smtp;open;Exim smtpd;;syn-ack;4.76;10
    127.0.0.1;tcp;53;domain;open;dnsmasq;;syn-ack;2.59;10
    127.0.0.1;tcp;80;http;open;Apache httpd;(Ubuntu);syn-ack;2.2.22;10
    127.0.0.1;tcp;111;rpcbind;open;;;syn-ack;;10
    127.0.0.1;tcp;139;netbios-ssn;open;Samba smbd;workgroup: WORKGROUP;syn-ack;3.X;10
    127.0.0.1;tcp;443;;open;;;syn-ack;;


To check the network status
---------------------------

.. code-block:: python

    >>> nm.scan(hosts='192.168.1.0/24', arguments='-n -sP -PE -PA21,23,80,3389')
    >>> hosts_list = [(x, nm[x]['status']['state']) for x in nm.all_hosts()]
    >>> for host, status in hosts_list:
    >>>     print('{0}:{1}'.host)
    192.168.1.0:down
    192.168.1.1:up
    192.168.1.10:down
    192.168.1.100:down
    192.168.1.101:down
    192.168.1.102:down
    192.168.1.103:down
    192.168.1.104:down
    192.168.1.105:down
    [...]


Using a Scanner Async
---------------------

.. code-block:: python

    >>> nma = nmap.PortScannerAsync()
    >>> def callback_result(host, scan_result):
    >>>     print '------------------'
    >>>     print host, scan_result
    >>>
    >>> nma.scan(hosts='192.168.1.0/30', arguments='-sP', callback=callback_result)
    >>> while nma.still_scanning():
    >>>     print("Waiting >>>")
    >>>     nma.wait(2)   # you can do whatever you want but I choose to wait after the end of the scan
    >>>
    192.168.1.1 {'nmap': {'scanstats': {'uphosts': '1', 'timestr': 'Mon Jun  7 11:31:11 2010', 'downhosts': '0', 'totalhosts': '1', 'elapsed': '0.43'}, 'scaninfo': {}, 'command_line': 'nmap -oX - -sP 192.168.1.1'}, 'scan': {'192.168.1.1': {'status': {'state': 'up', 'reason': 'arp-response'}, 'hostname': 'neufbox'}}}
    ------------------
    192.168.1.2 {'nmap': {'scanstats': {'uphosts': '0', 'timestr': 'Mon Jun  7 11:31:11 2010', 'downhosts': '1', 'totalhosts': '1', 'elapsed': '0.29'}, 'scaninfo': {}, 'command_line': 'nmap -oX - -sP 192.168.1.2'}, 'scan': {'192.168.1.2': {'status': {'state': 'down', 'reason': 'no-response'}, 'hostname': ''}}}
    ------------------
    192.168.1.3 {'nmap': {'scanstats': {'uphosts': '0', 'timestr': 'Mon Jun  7 11:31:11 2010', 'downhosts': '1', 'totalhosts': '1', 'elapsed': '0.29'}, 'scaninfo': {}, 'command_line': 'nmap -oX - -sP 192.168.1.3'}, 'scan': {'192.168.1.3': {'status': {'state': 'down', 'reason': 'no-response'}, 'hostname': ''}}}
    >>> nm = nmap.PortScannerYield() >>> for progressive_result in nm.scan('127.0.0.1/24', '22-25'): >>> print(progressive_result)


See also example.py in archive file.


Using a Scanner Async
---------------------

.. code-block:: python

    >>> nm = nmap.PortScanner()                                                                                                                                                                                                              
    >>> nm.scan('127.0.0.1', '22-40043', timeout=10)
    PortScannerTimeout: 'Timeout from nmap process'

Contributors
============

.. code-block:: text

    Steve 'Ashcrow' Milner
    Brian Bustin
    old.schepperhand
    Johan Lundberg
    Thomas D. maaaaz
    Robert Bost
    David Peltier
    Ed Jones


Homepage
========

http://xael.org/norman/python/python-nmap/


.. |PyPI Version| image:: https://img.shields.io/pypi/pyversions/python-nmap.svg?maxAge=2592000
   :target: https://pypi.python.org/pypi/python-nmap
.. |PyPI License| image:: https://img.shields.io/pypi/l/python-nmap.svg?maxAge=2592000
   :target: https://pypi.python.org/pypi/python-nmap
.. |PyPI latest| image:: https://img.shields.io/pypi/v/python-nmap.svg?maxAge=360
   :target: https://pypi.python.org/pypi/python-nmap


CHANGELOG
Changelog
=========

2021/06/29 (v0.7.0.2)
---------------------

   - Drop universal wheel


2021/06/29 (v0.7.0.1)
-------------------

   - Fix building Python Wheels
   - Removes long description from setup

2020/03/02 (v0.7.0)
-------------------

   - Add black and flake8 for development
   - Drop support for python2.7

2020/02/28 (v0.6.4)
-------------------

  - Add timeout parameter and PortScannerTimeout exception

2018/09/23 (v0.6.3)
-------------------

  - Refactor Readme and changelog files for the Pypi pattern and applying styles to the blocks for better visualization
  - Change setup.py to use an explicit version of python, since asynchronous tasks are not available for python 2.x
  - Add clean method to the Makefile
  - Update Manifest file
  - Update version and Fix PEP8 on nmap/nmap.py

2017/01/07 (v0.6.2)
-------------------


2016/07/29 (v0.6.1)
-------------------

  - Fix bug #22 UnboundLocalError in scan_progressive
  - Fix bug #23 Scanning fails on nmap warnings
  - Fix bug #20: Fix for empty <hostnames> values which results in blank CSV output
  - Fix bug #18: print(nm.csv()) does not return any results
  - Fix bug #19: nmap program was not found in path
  - Fix bug #17: hostname is no longer reported


2016/03/15 (v0.6.0)
-------------------

 - Add information about Nmap special licence
 - Licence precision for distributing python-nmap along with nmap

2016/03/15 (v0.5.2)
-------------------

 - add hostname to csv export

2015/12/05 (v0.5.0-1)
---------------------

 - updating example.py

2015/11/18 (v0.5.0)
-------------------

 - Closes bugs :
   - #11 Display only one osclass/osmatch instead of multiple

 - Change in data structure :
   - osmatch is a list of osclass
   - osclass is a list of dictionnary
   - added cpe which is a list of string
   - added portused which is a list of dictionnary

   Data structure for a host looks like :

.. code-block:: python

  {'addresses': {'ipv4': '127.0.0.1'},
    'hostnames': [],
    'osmatch': [{'accuracy': '98',
                'line': '36241',
                'name': 'Juniper SA4000 SSL VPN gateway (IVE OS 7.0)',
                'osclass': [{'accuracy': '98',
                              'cpe': ['cpe:/h:juniper:sa4000',
                                      'cpe:/o:juniper:ive_os:7'],
                              'osfamily': 'IVE OS',
                              'osgen': '7.X',
                              'type': 'firewall',
                              'vendor': 'Juniper'}]},
                {'accuracy': '91',
                'line': '17374',
                'name': 'Citrix Access Gateway VPN gateway',
                'osclass': [{'accuracy': '91',
                              'cpe': [],
                              'osfamily': 'embedded',
                              'osgen': None,
                              'type': 'proxy server',
                              'vendor': 'Citrix'}]}],
    'portused': [{'portid': '443', 'proto': 'tcp', 'state': 'open'},
                {'portid': '113', 'proto': 'tcp', 'state': 'closed'}],
    'status': {'reason': 'syn-ack', 'state': 'up'},
    'tcp': {113: {'conf': '3',
                  'cpe': '',
                  'extrainfo': '',
                  'name': 'ident',
                  'product': '',
                  'reason': 'conn-refused',
                  'state': 'closed',
                  'version': ''},
            443: {'conf': '10',
                  'cpe': '',
                  'extrainfo': '',
                  'name': 'http',
                  'product': 'Juniper SA2000 or SA4000 VPN gateway http config',
                  'reason': 'syn-ack',
                  'state': 'open',
                  'version': ''}},
    'vendor': {}}


2015/11/17 (v0.4.7)
-------------------

  - Closes bugs:
    - #10 Error when trying to parse 'osclass' , 'osmatch'
    removed addresses, hostnames, status, vendor, osclass, uptime, osmatch
    from all_protocols()
  - Changed shebang line from python3 to python as it works with python2

2015/11/13 (v0.4.6)
-------------------

  - Closes bugs :
    - #10 Error when trying to parse 'osclass' , 'osmatch'

2015/10/25 (v0.4.5)
-------------------

  - Closes bugs :
    - #9 Can not pass ports with unicode string at scan function

2015/10/17 (v0.4.4)
-------------------

  - Closes bugs :
    - #8 IPv6 Async scanner doesn't work

2015/09/11 (v0.4.3)
-------------------

  - Change in url for __get_last_online_version

2015/09/11 (v0.4.2)
-------------------

  - Closes bugs :
    - #7: Error with empty hostname
    - #6: Windows support of close_fds if you redirect stdin/stdout/stderr

2015/08/21 (v0.4.1)
-------------------

  - Closes bugs :
    - #5: only one hostname stored per host
  - Add hostnames() method which return the list of hostnames as a list of
    dict [{'name':'hostname1', 'type':'PTR'}, {'name':'hostname2', 'type':'user'}]

2015/08/01 (v0.4.0)
-------------------

  - Closes bugs :
    - #2: use close_fds in subprocess.Popen
    - #3: memory leak parsing xml using xml.dom.minidom
  - Corrects a bug in parsing osclass
  - Add nosetests for case testing
  - Removed test case in docstring

2015/05/08 (v0.3.7)
-------------------

  - adding sudo parameter for scanning (idea from scupython)

2015/05/08 (v0.3.6)
-------------------

  - correcting issue 7 : Issues under windows

2015/05/08 (v0.3.5)
-------------------

  - correcting a bug in all_protocols()
  - correcting issue 8 : PortScannerAsync Doesn't work in windows...

2014/06/22 (v0.3.4)
-------------------

  - adding PortScannerYield class with generator
    >>> nm = nmap.PortScannerYield()
    >>> for i in nm.scan('127.0.0.1/24', '22-25'):
    >>>     print(i)

2014/03/13 (v0.3.3)
-------------------

 - moving file example.py
 - adding function convert_nmap_output_to_encoding
 - adding vendor for mac address

2013/09/23 (v0.3.2)
-------------------

 - adding acces to CPE values under [host][proto][port]['cpe'] key

2013/07/27 (v0.3.1)

nmap/__init__.py
# -*- coding: latin-1 -*-

"""
nmap.py - v0.2.1 - 2010.12.15

Author : Alexandre Norman - norman@xael.org
Contributors: Steve 'Ashcrow' Milner - steve@gnulinux.net
              Brian Bustin - brian at bustin.us
Licence : GPL v3 or any later version


This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


Test strings :
^^^^^^^^^^^^
>>> import nmap
>>> nm = nmap.PortScanner()
>>> nm.scan('127.0.0.1', '22-443')
>>> nm.command_line()
'nmap -oX - -p 22-443 -sV 127.0.0.1'
>>> nm.scaninfo()
{'tcp': {'services': '22-443', 'method': 'connect'}}
>>> nm.all_hosts()
['127.0.0.1']
>>> nm['127.0.0.1'].hostname()
'localhost'
>>> nm['127.0.0.1'].state()
'up'
>>> nm['127.0.0.1'].all_protocols()
['tcp']
>>> nm['127.0.0.1']['tcp'].keys()
[80, 25, 443, 22, 111]
>>> nm['127.0.0.1'].has_tcp(22)
True
>>> nm['127.0.0.1'].has_tcp(23)
False
>>> nm['127.0.0.1']['tcp'][22]
{'state': 'open', 'reason': 'syn-ack', 'name': 'ssh'}
>>> nm['127.0.0.1'].tcp(22)
{'state': 'open', 'reason': 'syn-ack', 'name': 'ssh'}
>>> nm['127.0.0.1']['tcp'][22]['state']
'open'
"""

from .nmap import *


example.py
#!/usr/bin/env python
# -*- coding: latin-1 -*-

"""
example.py - 2011.11.09

Author : Alexandre Norman - norman@xael.org
Contributor: Steve 'Ashcrow' Milner - steve@gnulinux.net
Licence : GPL v3 or any later version


This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""
import sys
import os

import nmap                         # import nmap.py module
try:
    nm = nmap.PortScanner()         # instantiate nmap.PortScanner object
except nmap.PortScannerError:
    print('Nmap not found', sys.exc_info()[0])
    sys.exit(1)
except:
    print("Unexpected error:", sys.exc_info()[0])
    sys.exit(1)


# Data structure looks like :
#
#      {'addresses': {'ipv4': '127.0.0.1'},
#       'hostnames': [],
#       'osmatch': [{'accuracy': '98',
#                    'line': '36241',
#                    'name': 'Juniper SA4000 SSL VPN gateway (IVE OS 7.0)',
#                    'osclass': [{'accuracy': '98',
#                                 'cpe': ['cpe:/h:juniper:sa4000',
#                                         'cpe:/o:juniper:ive_os:7'],
#                                 'osfamily': 'IVE OS',
#                                 'osgen': '7.X',
#                                 'type': 'firewall',
#                                 'vendor': 'Juniper'}]},
#                   {'accuracy': '91',
#                    'line': '17374',
#                    'name': 'Citrix Access Gateway VPN gateway',
#                    'osclass': [{'accuracy': '91',
#                                 'cpe': [],
#                                 'osfamily': 'embedded',
#                                 'osgen': None,
#                                 'type': 'proxy server',
#                                 'vendor': 'Citrix'}]}],
#       'portused': [{'portid': '443', 'proto': 'tcp', 'state': 'open'},
#                    {'portid': '113', 'proto': 'tcp', 'state': 'closed'}],
#       'status': {'reason': 'syn-ack', 'state': 'up'},
#       'tcp': {113: {'conf': '3',
#                     'cpe': '',
#                     'extrainfo': '',
#                     'name': 'ident',
#                     'product': '',
#                     'reason': 'conn-refused',
#                     'state': 'closed',
#                     'version': ''},
#               443: {'conf': '10',
#                     'cpe': '',
#                     'extrainfo': '',
#                     'name': 'http',
#                     'product': 'Juniper SA2000 or SA4000 VPN gateway http config',
#                     'reason': 'syn-ack',
#                     'state': 'open',
#                     'version': ''}},
#       'vendor': {}}



nm.scan('127.0.0.1', '22-443')      # scan host 127.0.0.1, ports from 22 to 443
nm.command_line()                   # get command line used for the scan : nmap -oX - -p 22-443 127.0.0.1
nm.scaninfo()                       # get nmap scan informations {'tcp': {'services': '22-443', 'method': 'connect'}}
nm.all_hosts()                      # get all hosts that were scanned
nm['127.0.0.1'].hostname()          # get one hostname for host 127.0.0.1, usualy the user record
nm['127.0.0.1'].hostnames()         # get list of hostnames for host 127.0.0.1 as a list of dict [{'name':'hostname1', 'type':'PTR'}, {'name':'hostname2', 'type':'user'}]
nm['127.0.0.1'].state()             # get state of host 127.0.0.1 (up|down|unknown|skipped) 
nm['127.0.0.1'].all_protocols()     # get all scanned protocols ['tcp', 'udp'] in (ip|tcp|udp|sctp)
if ('tcp' in nm['127.0.0.1']):
    list(nm['127.0.0.1']['tcp'].keys()) # get all ports for tcp protocol

nm['127.0.0.1'].all_tcp()           # get all ports for tcp protocol (sorted version)
nm['127.0.0.1'].all_udp()           # get all ports for udp protocol (sorted version)
nm['127.0.0.1'].all_ip()            # get all ports for ip protocol (sorted version)
nm['127.0.0.1'].all_sctp()          # get all ports for sctp protocol (sorted version)
if nm['127.0.0.1'].has_tcp(22):     # is there any information for port 22/tcp on host 127.0.0.1
    nm['127.0.0.1']['tcp'][22]          # get infos about port 22 in tcp on host 127.0.0.1
    nm['127.0.0.1'].tcp(22)             # get infos about port 22 in tcp on host 127.0.0.1
    nm['127.0.0.1']['tcp'][22]['state'] # get state of port 22/tcp on host 127.0.0.1 (open


# a more usefull example :
for host in nm.all_hosts():
    print('----------------------------------------------------')
    print('Host : {0} ({1})'.format(host, nm[host].hostname()))
    print('State : {0}'.format(nm[host].state()))

    for proto in nm[host].all_protocols():
        print('----------')
        print('Protocol : {0}'.format(proto))

        lport = list(nm[host][proto].keys())
        lport.sort()
        for port in lport:
            print('port : {0}\tstate : {1}'.format(port, nm[host][proto][port]))


print('----------------------------------------------------')
# print result as CSV
print(nm.csv())



print('----------------------------------------------------')
# If you want to do a pingsweep on network 192.168.1.0/24:
nm.scan(hosts='192.168.0.0/24', arguments='-n -sP -PE -PA21,23,80,3389')
hosts_list = [(x, nm[x]['status']['state']) for x in nm.all_hosts()]
for host, status in hosts_list:
    print('{0}:{1}'.format(host, status))





print('----------------------------------------------------')
# Asynchronous usage of PortScannerAsync


nma = nmap.PortScannerAsync()

def callback_result(host, scan_result):
    print('------------------')
    print(host, scan_result)

nma.scan(hosts='192.168.0.0/30', arguments='-sP', callback=callback_result)

while nma.still_scanning():
    print("Waiting ...")
    nma.wait(2)   # you can do whatever you want but I choose to wait after the end of the scan

if (os.getuid() == 0):
    print('----------------------------------------------------')
    # Os detection (need root privileges)
    nm.scan("127.0.0.1", arguments="-O")
    if 'osmatch' in nm['127.0.0.1']:
        for osmatch in nm['127.0.0.1']['osmatch']:
            print('OsMatch.name : {0}'.format(osmatch['name']))
            print('OsMatch.accuracy : {0}'.format(osmatch['accuracy']))
            print('OsMatch.line : {0}'.format(osmatch['line']))
            print('')

            if 'osclass' in osmatch:
                for osclass in osmatch['osclass']:
                    print('OsClass.type : {0}'.format(osclass['type']))
                    print('OsClass.vendor : {0}'.format(osclass['vendor']))
                    print('OsClass.osfamily : {0}'.format(osclass['osfamily']))
                    print('OsClass.osgen : {0}'.format(osclass['osgen']))
                    print('OsClass.accuracy : {0}'.format(osclass['accuracy']))
                    print('')


    if 'fingerprint' in nm['127.0.0.1']:
        print('Fingerprint : {0}'.format(nm['127.0.0.1']['fingerprint']))


    # Vendor list for MAC address
    print('scanning localnet')
    nm.scan('192.168.0.0/24', arguments='-O')
    for h in nm.all_hosts():
        print(h)
        if 'mac' in nm[h]['addresses']:
            print(nm[h]['addresses'], nm[h]['vendor'])



print('----------------------------------------------------')
# Read output captured to a file
# Example : nmap -oX - -p 22-443 -sV 127.0.0.1 > nmap_output.xml

with open("./nmap_output.xml", "r") as fd:
    content = fd.read()
    nm.analyse_nmap_xml_scan(content)
    print(nm.csv())



print('----------------------------------------------------')
# Progressive scan with generator
nm = nmap.PortScannerYield()
for progressive_result in nm.scan('127.0.0.1/24', '22-25'):
    print(progressive_result)


print('----------------------------------------------------')
# Using a timeout
nm = nmap.PortScanner()
nm.scan('127.0.0.1', '22-40043', timeout=1)
